src/cluster/backend/zeromq/ZeroMQ.cc

```diff
@@ -0,0 +1,526 @@
+        std::string name;
+        std::function<void(std::vector<MultipartMessage>&)> handler;
+    };
+
+    std::vector<SocketInfo> sockets = {
+        {.socket = child_inproc, .name = "inproc", .handler = HandleInprocMessages},
+        {.socket = xpub, .name = "xpub", .handler = HandleXPubMessages},
+        {.socket = xsub, .name = "xsub", .handler = HandleXSubMessages},
+        {.socket = log_pull, .name = "log_pull", .handler = HandleLogMessages},
+    };
+
+    std::vector<zmq::pollitem_t> poll_items(sockets.size());
+
+    while ( true ) {
+        for ( size_t i = 0; i < sockets.size(); i++ )
+            poll_items[i] = {.socket = sockets[i].socket.handle(), .fd = 0, .events = ZMQ_POLLIN | ZMQ_POLLERR};
+
+        // Awkward.
+        std::vector<std::vector<MultipartMessage>> rcv_messages(sockets.size());
+        try {
+            int r = zmq::poll(poll_items, std::chrono::seconds(-1));
+            ZEROMQ_DEBUG_THREAD_PRINTF(DebugFlag::POLL, "poll: r=%d", r);
+
+            for ( size_t i = 0; i < poll_items.size(); i++ ) {
+                const auto& item = poll_items[i];
+                ZEROMQ_DEBUG_THREAD_PRINTF(DebugFlag::POLL, "poll: items[%lu]=%s %s %s\n", i, sockets[i].name.c_str(),
+                                           item.revents & ZMQ_POLLIN ? "pollin " : "",
+                                           item.revents & ZMQ_POLLERR ? "err" : "");
+
+                if ( item.revents & ZMQ_POLLERR ) {
+                    // What should we be doing? Re-open sockets? Terminate?
+                    ZEROMQ_THREAD_PRINTF("poll: error: POLLERR on socket %zu %s %p revents=%x\n", i,
+                                         sockets[i].name.c_str(), item.socket, item.revents);
+                }
+
+                // Nothing to do?
+                if ( (item.revents & ZMQ_POLLIN) == 0 )
+                    continue;
+
+                bool consumed_one = false;
+
+                // Read as many messages as possible.
+                do {
+                    zmq::message_t msg;
+                    rcv_messages[i].emplace_back(); // make room for a multiparte message
+
+                    auto& into = rcv_messages[i].back();
+
+                    consumed_one = false;
+                    bool more = false;
+
+                    // Read a multi-part message.
+                    do {
+                        auto r = sockets[i].socket.recv(msg, zmq::recv_flags::dontwait);
+                        if ( r ) {
+                            consumed_one = true;
+                            more = msg.more();
+                            into.emplace_back(std::move(msg));
+                        }
+                        else {
+                            // EAGAIN and more flag set? Try again, otherwise.
+                            if ( more )
+                                continue;
+                        }
+                    } while ( more );
+                } while ( consumed_one );
```

Comment:
Yeah, could be a concern. Without seeing it in it's a bit hand-wavy though.

I do agree we could yield here when reaching a certain number of messages from a socket. Is that what you were thinking?

Reply:
I added an option that might help to avoid this a bit, though I'd postpone to real deployment to see what falls apart :-)

Reply:
Yeah, I think that would be enough to start with. For example the poll code in the main loop only polls every X number of loops, to avoid this exact issue.